### ISSUE TITLE
Improve policy performance with longer training runs

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/bucket-brigade#10]
## Goal
Improve the trained policy performance by running longer training sessions.

## Background
Initial training runs used only 50K steps, which resulted in:
- Mean reward: -79.82 ± 177.70
- High variance in episode performance
- Inconsistent cooperation behavior

Longer training (500K-1M steps) should allow the policy to:
- Learn better coordination strategies
- Reduce variance in performance
- Achieve more consistent positive rewards

## Tasks
- [ ] Run 500K step training with default parameters
- [ ] Run 1M step training with default parameters
- [ ] Compare performance metrics across different training lengths
- [ ] Document optimal training duration for this environment
- [ ] Save best performing checkpoints

## Commands
```bash
# 500K steps
uv run python scripts/train_simple.py \
    --num-steps 500000 \
    --num-opponents 3 \
    --batch-size 2048 \
    --hidden-size 128 \
    --save-path models/policy_500k.pt

# 1M steps
uv run python scripts/train_simple.py \
    --num-steps 1000000 \
    --num-opponents 3 \
    --batch-size 2048 \
    --hidden-size 128 \
    --save-path models/policy_1m.pt

# Evaluate
uv run python scripts/evaluate_simple.py \
    --model-path models/policy_1m.pt \
    --num-episodes 100
```

## Success Criteria
- Mean reward > 0 (positive net benefit)
- Standard deviation < 100 (more consistent)
- At least 70% of episodes have positive rewards

## Related
See TRAINING_GUIDE.md for more details.

**Comment by rjwalters** [comment]
## Curator Enhancement

### Test Plan

Before marking this complete, verify:

- [ ] **Training runs complete successfully**
  - [ ] 500K step training completes without errors
  - [ ] 1M step training completes without errors
  - [ ] Checkpoints saved correctly at specified paths
  
- [ ] **Performance metrics meet criteria**
  - [ ] Mean reward > 0 (currently -79.82)
  - [ ] Standard deviation < 100 (currently 177.70)
  - [ ] At least 70% episodes have positive rewards
  
- [ ] **Evaluation runs successfully**
  - [ ] Evaluate script runs on both saved models
  - [ ] 100 episodes complete for statistical validity
  - [ ] Results logged and compared

- [ ] **Documentation updated**
  - [ ] TRAINING_GUIDE.md updated with optimal duration findings
  - [ ] Performance comparison table added
  - [ ] Best checkpoint locations documented

### Implementation Notes

**Resource Requirements:**
- Expect 500K steps to take ~2-4 hours on CPU (varies by hardware)
- 1M steps may take 4-8 hours
- Models will be ~500KB-2MB each (depending on hidden size)
- Consider using tmux/screen for long-running training

**Monitoring During Training:**
- Training script should output periodic updates
- Watch for reward trends (should improve over time)
- Monitor for crashes or OOM errors
- Check GPU/CPU utilization if available

**Comparison Strategy:**
- Save training logs for all runs
- Plot reward curves: 50K vs 500K vs 1M
- Compare final evaluation metrics side-by-side
- Identify diminishing returns point (if 500K ≈ 1M, prefer 500K)

**Related Issues:**
- Consider #11 (hyperparameter tuning) after establishing baseline
- Results may inform #12 (scenario-specific training)
- May benefit from #13 (TensorBoard logging) for better visualization

### Dependencies

None identified - this is a standalone training experiment.

---
📝 **Curator**: Added test plan, implementation notes, and resource requirements. Issue is well-formed and ready for implementation after human approval.

**Comment by rjwalters** [comment]
📌 **Dependency Note**: Issue #11 (hyperparameter tuning) is blocked waiting for this baseline to be established. Complete this issue first to enable effective hyperparameter comparison.

**Comment by rjwalters** [comment]
## Blocked: Requires GPU for Training

After attempting to start training, confirmed that this work requires GPU hardware to complete in a reasonable timeframe. CPU training would take 8+ hours per run.

**Next Steps:**
- Wait for GPU access before attempting this issue
- Consider cloud GPU instances (AWS, GCP, Lambda Labs) if local GPU unavailable
- Estimated GPU time: 30-60 minutes per 500K run on modern GPU

Marking as `loom:blocked` until GPU resources are available.

**Comment by rjwalters** [comment]
## Performance Optimization Update - GPU Training Throughput

Implemented and tested GPU-accelerated vectorized training optimizations.

### Current Performance

**Achieved Throughput**: ~8,700 steps/second
- Configuration: 128 parallel environments, 16K batch size, multiprocessing vectorization
- Hardware: Single NVIDIA L4 GPU (4 available)
- Benchmark: 10M steps completing in ~19 minutes

### Performance Comparison

| Configuration | Steps | Time | Throughput |
|--------------|-------|------|------------|
| Original CPU (est) | 1M | ~hours | ~100 steps/sec |
| Basic Vectorized | 500K | 0.9 min | ~9.3K steps/sec |
| Basic Vectorized | 1M | 1.8 min | ~9.3K steps/sec |
| Current Benchmark | 10M | ~19 min | **~8.7K steps/sec** |

### Bottleneck Analysis

**Current limiting factors**:
1. **Single GPU utilization**: Only GPU 0 is used (~3-15% util), 3 GPUs idle
2. **CPU-bound environment simulation**: Multiprocessing workers run game logic on CPU
3. **Batch size limitations**: Larger batches (>16K) cause initialization hangs
4. **Multiprocessing overhead**: 512+ environments deadlock during creation

### Recommendations for Higher Throughput

To achieve **millions of steps/second**, we need:

1. **Distributed training across all 4 GPUs**:
   - Use PyTorch DistributedDataParallel
   - Each GPU handles 32-64 environments
   - Could achieve 4x throughput: **~35K steps/sec**

2. **GPU-accelerated environments** (longer-term):
   - Port game logic to GPU kernels (CUDA/Jax)
   - Batch all environment operations on GPU
   - Potential 100x speedup: **~800K steps/sec**

3. **Optimized batch processing**:
   - Use metta's architecture: 256K-512K rollout batches
   - 4K-8K GPU minibatches
   - Requires fixing multiprocessing initialization issues

### Current Status

- ✅ Vectorized training working reliably at 8.7K steps/sec
- ✅ Can train 10M steps in under 20 minutes  
- ✅ Models achieving positive rewards (40-56 average)
- ⚠️  Using only 1 of 4 available GPUs
- ⚠️  Environments run on CPU (expected for current architecture)

The current implementation provides a **~87x speedup** over sequential training, making long training runs practical. Further optimization would require architectural changes for multi-GPU or GPU-native environments.
